File: venus_evcharger/dbus_adapter/read/pv_dormancy.py
# SPDX-License-Identifier: GPL-3.0-or-later
"""Bounded monotonic availability evidence for AC and DC PV sources."""

from __future__ import annotations

import math
import re
import time
from collections.abc import Callable
from dataclasses import dataclass
from typing import Literal, TypedDict

PvDormancyReason = Literal["explicit-dormant-state"]
PvSourceFailureReason = Literal["source-path-unreadable"]

DEFAULT_EVIDENCE_TTL_SECONDS = 18.0 * 60.0 * 60.0
DEFAULT_OBSERVATION_RETENTION_SECONDS = 24.0 * 60.0 * 60.0
DEFAULT_MAX_OBSERVATIONS = 64
DEFAULT_ERROR_BACKOFF_SECONDS = 300.0

_DORMANT_WORDS = frozenset({"asleep", "dormant", "sleeping", "standby"})
_NEGATED_DORMANT_PHRASES = (
    "not asleep",
    "not dormant",
    "not sleeping",
    "not in standby",
)


class PvDormancyEvidencePayload(TypedDict):
    """Serialized adapter-health representation of dormant-source evidence."""

    source_id: str
    reason: PvDormancyReason
    observed_at: float


@dataclass(frozen=True, slots=True)
class PvDormancyEvidence:
    """Semantic evidence exported to adapter health without DBus identities."""

    source_id: str
    reason: PvDormancyReason
    observed_at: float

    def to_payload(self) -> PvDormancyEvidencePayload:
        return {
            "source_id": self.source_id,
            "reason": self.reason,
            "observed_at": self.observed_at,
        }


@dataclass(frozen=True, slots=True)
class PvDormancyPolicy:
    """Bound memory, evidence lifetime, and retry pressure for PV sources."""

    evidence_ttl_seconds: float = DEFAULT_EVIDENCE_TTL_SECONDS
    observation_retention_seconds: float = DEFAULT_OBSERVATION_RETENTION_SECONDS
    max_observations: int = DEFAULT_MAX_OBSERVATIONS
    error_backoff_seconds: float = DEFAULT_ERROR_BACKOFF_SECONDS


DEFAULT_PV_DORMANCY_POLICY = PvDormancyPolicy()


@dataclass(slots=True)
class _PvObservation:
    validated: bool = False
    available: bool | None = None
    failure_reason: PvSourceFailureReason | None = None
    dormant_reason: PvDormancyReason | None = None
    dormant_observed_at: float = 0.0
    dormant_observed_monotonic: float = 0.0
    last_observed_monotonic: float = 0.0
    next_probe_monotonic: float = 0.0

    def semantic_state(
        self,
    ) -> tuple[
        bool,
        bool | None,
        PvSourceFailureReason | None,
        PvDormancyReason | None,
    ]:
        return (
            self.validated,
            self.available,
            self.failure_reason,
            self.dormant_reason,
        )


class PvDormancyTracker:
    """Track validated PV paths, bounded backoff, and explicit dormancy."""

    def __init__(
        self,
        *,
        policy: PvDormancyPolicy = DEFAULT_PV_DORMANCY_POLICY,
        monotonic: Callable[[], float] | None = None,
        wall_clock: Callable[[], float] | None = None,
    ) -> None:
        self._evidence_ttl_seconds = max(1.0, float(policy.evidence_ttl_seconds))
        self._observation_retention_seconds = max(
            self._evidence_ttl_seconds,
            float(policy.observation_retention_seconds),
        )
        self._max_observations = max(1, int(policy.max_observations))
        self._error_backoff_seconds = max(
            0.0,
            float(policy.error_backoff_seconds),
        )
        self._monotonic = time.monotonic if monotonic is None else monotonic
        self._wall_clock = time.time if wall_clock is None else wall_clock
        self._observations: dict[str, _PvObservation] = {}
        self._revision = 0
        self._evidence_cache_key: tuple[int, frozenset[str]] | None = None
        self._evidence_cache: tuple[PvDormancyEvidence, ...] = ()

    @property
    def observation_count(self) -> int:
        return len(self._observations)

    @property
    def revision(self) -> int:
        return self._revision

    def record_value(
        self,
        source_id: str,
        value: object,
        *,
        active_source_ids: frozenset[str] = frozenset(),
    ) -> bool:
        """Record one successful numeric sample and validate its PV path."""
        if _finite_number(value) is None:
            return False
        current = self._monotonic()
        observation = self._observation(source_id, active_source_ids)
        before = observation.semantic_state()
        observation.validated = True
        observation.available = True
        observation.failure_reason = None
        observation.dormant_reason = None
        observation.last_observed_monotonic = current
        observation.next_probe_monotonic = 0.0
        return self._record_change(observation.semantic_state() != before)

    def record_error(
        self,
        source_id: str,
        error: BaseException | str,
        *,
        active_source_ids: frozenset[str] = frozenset(),
    ) -> bool:
        """Record one failed sample without mistaking generic failure for sleep."""
        current = self._monotonic()
        observation = self._observation(source_id, active_source_ids)
        before = observation.semantic_state()
        observation.available = False
        observation.last_observed_monotonic = current
        observation.next_probe_monotonic = current + self._error_backoff_seconds
        if explicit_dormancy_error(error):
            observation.validated = True
            observation.failure_reason = None
            observation.dormant_reason = "explicit-dormant-state"
            observation.dormant_observed_at = max(0.0, float(self._wall_clock()))
            observation.dormant_observed_monotonic = current
        elif observation.dormant_reason is None:
            observation.failure_reason = "source-path-unreadable"
        return self._record_change(observation.semantic_state() != before)

    def maintain(self, active_source_ids: frozenset[str]) -> bool:
        """Expire semantic evidence and prune inactive history deterministically."""
        current = self._monotonic()
        changed = self._expire_dormancy(current)
        changed = self._prune_inactive(active_source_ids, current) or changed
        return self._record_change(changed)

    def probe_allowed(self, source_id: str) -> bool:
        observation = self._observations.get(source_id)
        return (
            observation is None
            or self._monotonic() >= observation.next_probe_monotonic
        )

    def source_validated(self, source_id: str) -> bool:
        observation = self._observations.get(source_id)
        return bool(observation is not None and observation.validated)

    def source_available(self, source_id: str) -> bool | None:
        observation = self._observations.get(source_id)
        return None if observation is None else observation.available

    def source_failure_reason(
        self,
        source_id: str,
    ) -> PvSourceFailureReason | None:
        observation = self._observations.get(source_id)
        return None if observation is None else observation.failure_reason

    def validated_source_ids(self) -> frozenset[str]:
        return frozenset(
            source_id
            for source_id, observation in self._observations.items()
            if observation.validated
        )

    def evidence(
        self,
        source_ids: frozenset[str],
    ) -> tuple[PvDormancyEvidence, ...]:
        cache_key = (self._revision, source_ids)
        if cache_key == self._evidence_cache_key:
            return self._evidence_cache
        candidates = (
            _observation_evidence(source_id, observation, source_ids)
            for source_id, observation in sorted(self._observations.items())
        )
        evidence = tuple(item for item in candidates if item is not None)
        self._evidence_cache_key = cache_key
        self._evidence_cache = evidence
        return evidence

    def _observation(
        self,
        source_id: str,
        active_source_ids: frozenset[str],
    ) -> _PvObservation:
        observation = self._observations.get(source_id)
        if observation is not None:
            return observation
        self._trim_to_capacity(active_source_ids, reserve=1)
        observation = _PvObservation()
        self._observations[source_id] = observation
        return observation

    def _expire_dormancy(self, current: float) -> bool:
        changed = False
        for observation in self._observations.values():
            if observation.dormant_reason is None:
                continue
            if (
                current - observation.dormant_observed_monotonic
                < self._evidence_ttl_seconds
            ):
                continue
            observation.dormant_reason = None
            observation.failure_reason = "source-path-unreadable"
            changed = True
        return changed

    def _prune_inactive(
        self,
        active_source_ids: frozenset[str],
        current: float,
    ) -> bool:
        expired = [
            source_id
            for source_id, observation in self._observations.items()
            if (
                source_id not in active_source_ids
                and current - observation.last_observed_monotonic
                >= self._observation_retention_seconds
            )
        ]
        for source_id in expired:
            del self._observations[source_id]
        return bool(expired)

    def _trim_to_capacity(
        self,
        active_source_ids: frozenset[str],
        *,
        reserve: int,
    ) -> None:
        target_size = max(0, self._max_observations - reserve)
        while len(self._observations) > target_size:
            source_id = min(
                self._observations,
                key=lambda candidate: (
                    candidate in active_source_ids,
                    self._observations[candidate].last_observed_monotonic,
                    candidate,
                ),
            )
            del self._observations[source_id]

    def _record_change(self, changed: bool) -> bool:
        if changed:
            self._revision += 1
        return changed


def explicit_dormancy_error(error: BaseException | str) -> bool:
    """Return true only for an explicit sleep/standby statement."""
    message = str(error).strip().lower()
    if not message or any(phrase in message for phrase in _NEGATED_DORMANT_PHRASES):
        return False
    return bool(_DORMANT_WORDS.intersection(re.findall(r"[a-z]+", message)))


def _observation_evidence(
    source_id: str,
    observation: _PvObservation,
    source_ids: frozenset[str],
) -> PvDormancyEvidence | None:
    if source_id not in source_ids:
        return None
    if not observation.validated:
        return None
    if observation.available is not False:
        return None
    if observation.dormant_reason is None:
        return None
    return PvDormancyEvidence(
        source_id=source_id,
        reason=observation.dormant_reason,
        observed_at=observation.dormant_observed_at,
    )


def _finite_number(value: object) -> float | None:
    if isinstance(value, bool) or not isinstance(value, int | float):
        return None
    numeric = float(value)
    return numeric if math.isfinite(numeric) else None


__all__ = [
    "PvDormancyEvidence",
    "PvDormancyEvidencePayload",
    "PvDormancyPolicy",
    "PvDormancyReason",
    "PvDormancyTracker",
    "PvSourceFailureReason",
    "explicit_dormancy_error",
]

File: rust/dbus-adapter/src/reader/pv_dormancy.rs
// SPDX-License-Identifier: GPL-3.0-or-later
//! Bounded monotonic availability evidence for AC and DC PV sources.

use std::collections::{HashMap, HashSet};
use std::time::Duration;

use serde::Serialize;

use crate::energy::Clocks;

const EVIDENCE_TTL: Duration = Duration::from_secs(18 * 60 * 60);
const OBSERVATION_RETENTION: Duration = Duration::from_secs(24 * 60 * 60);
const ERROR_BACKOFF: Duration = Duration::from_secs(300);
const MAX_OBSERVATIONS: usize = 64;

#[derive(Clone, Debug, Serialize)]
pub(super) struct PvDormancyEvidence {
    pub(super) source_id: String,
    pub(super) reason: &'static str,
    pub(super) observed_at: f64,
}

#[derive(Clone, Debug, Default)]
struct PvObservation {
    validated: bool,
    available: Option<bool>,
    failure_reason: Option<&'static str>,
    dormant: bool,
    dormant_observed_at: f64,
    dormant_observed_monotonic: f64,
    last_observed_monotonic: f64,
    next_probe_monotonic: f64,
}

#[derive(Default)]
pub(super) struct PvDormancyTracker {
    observations: HashMap<String, PvObservation>,
}

impl PvDormancyTracker {
    pub(super) fn record_value(
        &mut self,
        source_id: &str,
        clocks: Clocks,
        active_source_ids: &HashSet<String>,
    ) {
        let observation = self.observation(source_id, active_source_ids);
        observation.validated = true;
        observation.available = Some(true);
        observation.failure_reason = None;
        observation.dormant = false;
        observation.last_observed_monotonic = clocks.monotonic;
        observation.next_probe_monotonic = 0.0;
    }

    pub(super) fn record_error(
        &mut self,
        source_id: &str,
        error: &str,
        clocks: Clocks,
        active_source_ids: &HashSet<String>,
    ) {
        let observation = self.observation(source_id, active_source_ids);
        observation.available = Some(false);
        observation.last_observed_monotonic = clocks.monotonic;
        observation.next_probe_monotonic = clocks.monotonic + ERROR_BACKOFF.as_secs_f64();
        if explicit_dormancy_error(error) {
            observation.validated = true;
            observation.failure_reason = None;
            observation.dormant = true;
            observation.dormant_observed_at = clocks.epoch.max(0.0);
            observation.dormant_observed_monotonic = clocks.monotonic;
        } else if !observation.dormant {
            observation.failure_reason = Some("source-path-unreadable");
        }
    }

    pub(super) fn maintain(&mut self, active_source_ids: &HashSet<String>, monotonic_at: f64) {
        for observation in self.observations.values_mut() {
            if observation.dormant
                && monotonic_at - observation.dormant_observed_monotonic
                    >= EVIDENCE_TTL.as_secs_f64()
            {
                observation.dormant = false;
                observation.failure_reason = Some("source-path-unreadable");
            }
        }
        self.observations.retain(|source_id, observation| {
            active_source_ids.contains(source_id)
                || monotonic_at - observation.last_observed_monotonic
                    < OBSERVATION_RETENTION.as_secs_f64()
        });
    }

    pub(super) fn probe_allowed(&self, source_id: &str, monotonic_at: f64) -> bool {
        self.observations
            .get(source_id)
            .is_none_or(|observation| monotonic_at >= observation.next_probe_monotonic)
    }

    pub(super) fn validated(&self, source_id: &str) -> bool {
        self.observations
            .get(source_id)
            .is_some_and(|observation| observation.validated)
    }

    pub(super) fn validated_source_ids(&self) -> HashSet<String> {
        self.observations
            .iter()
            .filter(|(_source_id, observation)| observation.validated)
            .map(|(source_id, _observation)| source_id.clone())
            .collect()
    }

    pub(super) fn evidence(&self, source_ids: &HashSet<String>) -> Vec<PvDormancyEvidence> {
        let mut result = self
            .observations
            .iter()
            .filter(|(source_id, observation)| {
                source_ids.contains(*source_id)
                    && observation.validated
                    && observation.available == Some(false)
                    && observation.dormant
            })
            .map(|(source_id, observation)| PvDormancyEvidence {
                source_id: source_id.clone(),
                reason: "explicit-dormant-state",
                observed_at: observation.dormant_observed_at,
            })
            .collect::<Vec<_>>();
        result.sort_by(|left, right| left.source_id.cmp(&right.source_id));
        result
    }

    pub(super) fn failure_reason(&self, source_id: &str) -> Option<&'static str> {
        self.observations
            .get(source_id)
            .and_then(|observation| observation.failure_reason)
    }

    fn observation<'a>(
        &'a mut self,
        source_id: &str,
        active_source_ids: &HashSet<String>,
    ) -> &'a mut PvObservation {
        if !self.observations.contains_key(source_id) {
            self.trim_to_capacity(active_source_ids, 1);
        }
        self.observations.entry(source_id.to_owned()).or_default()
    }

    fn trim_to_capacity(&mut self, active_source_ids: &HashSet<String>, reserve: usize) {
        let target = MAX_OBSERVATIONS.saturating_sub(reserve);
        while self.observations.len() > target {
            let candidate = self
                .observations
                .iter()
                .min_by(|left, right| {
                    (
                        active_source_ids.contains(left.0),
                        ordered_time(left.1.last_observed_monotonic),
                        left.0,
                    )
                        .cmp(&(
                            active_source_ids.contains(right.0),
                            ordered_time(right.1.last_observed_monotonic),
                            right.0,
                        ))
                })
                .map(|(source_id, _observation)| source_id.clone());
            let Some(source_id) = candidate else {
                break;
            };
            self.observations.remove(&source_id);
        }
    }
}

fn explicit_dormancy_error(error: &str) -> bool {
    let message = error.trim().to_ascii_lowercase();
    if message.is_empty()
        || [
            "not asleep",
            "not dormant",
            "not sleeping",
            "not in standby",
        ]
        .iter()
        .any(|phrase| message.contains(phrase))
    {
        return false;
    }
    message
        .split(|character: char| !character.is_ascii_alphabetic())
        .any(|word| matches!(word, "asleep" | "dormant" | "sleeping" | "standby"))
}

const fn ordered_time(value: f64) -> u64 {
    value.max(0.0).to_bits()
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    use super::{PvDormancyTracker, explicit_dormancy_error};
    use crate::energy::Clocks;
    use crate::runtime_policy_contract;

    #[test]
    fn only_explicit_non_negated_sleep_messages_are_dormant() {
        assert!(explicit_dormancy_error("inverter sleeping"));
        assert!(explicit_dormancy_error("DC inverter standby"));
        assert!(!explicit_dormancy_error("not sleeping"));
        assert!(!explicit_dormancy_error(
            "org.freedesktop.DBus.Error.NoReply"
        ));
    }

    #[test]
    fn generated_python_dormancy_messages_match() -> Result<(), String> {
        for case in runtime_policy_contract::load()?.dormancy_messages {
            assert_eq!(explicit_dormancy_error(&case.message), case.explicit);
        }
        Ok(())
    }

    #[test]
    fn generic_failure_backs_off_without_creating_dormancy() {
        let active = HashSet::from(["pv-a".to_owned()]);
        let mut tracker = PvDormancyTracker::default();
        tracker.record_error(
            "pv-a",
            "NoReply",
            Clocks {
                epoch: 10.0,
                monotonic: 20.0,
            },
            &active,
        );
        assert!(!tracker.probe_allowed("pv-a", 319.9));
        assert!(tracker.probe_allowed("pv-a", 320.0));
        assert!(tracker.evidence(&active).is_empty());
    }

    #[test]
    fn explicit_sleep_is_validated_evidence_and_success_clears_it() {
        let active = HashSet::from(["pv-a".to_owned()]);
        let mut tracker = PvDormancyTracker::default();
        tracker.record_error(
            "pv-a",
            "inverter asleep",
            Clocks {
                epoch: 10.0,
                monotonic: 20.0,
            },
            &active,
        );
        assert!(tracker.validated("pv-a"));
        assert_eq!(tracker.evidence(&active).len(), 1);
        tracker.record_value(
            "pv-a",
            Clocks {
                epoch: 30.0,
                monotonic: 320.0,
            },
            &active,
        );
        assert!(tracker.evidence(&active).is_empty());
    }
}
